sdks/java/extensions/euphoria/euphoria-core/src/main/java/org/apache/beam/sdk/extensions/euphoria/core/translate/coder/ClassAwareKryoCoder.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.euphoria.core.translate.coder;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.esotericsoftware.kryo.io.Output;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Objects;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.coders.CoderException;
+import org.apache.beam.sdk.extensions.euphoria.core.client.functional.VoidFunction;
+import org.objenesis.strategy.StdInstantiatorStrategy;
+
+/**
+ * {@link Coder} which encodes/decodes objects of given {@link Class} through {@link Kryo}.
+ */
+public class ClassAwareKryoCoder<T> extends Coder<T> {
+
+  /**
+   * Initial size of byte buffers in {@link Output}, {@link Input}.
+   */
+  private static final int DEFAULT_BUFFER_SIZE = 4096;
+
+  // this factory needs to be serializable, since the Coder itself is
+  private static final VoidFunction<Kryo> KRYO_FACTORY =
+      () -> {
+        final Kryo instance = new Kryo();
+        ((Kryo.DefaultInstantiatorStrategy) instance.getInstantiatorStrategy())
+            .setFallbackInstantiatorStrategy(new StdInstantiatorStrategy());
+        return instance;
+      };
+
+  private transient Kryo kryo;
+  private transient Output output;
+  private transient Input input;
+
+  private Class<T> clazz;
+
+  public ClassAwareKryoCoder(Class<T> clazz) {
+    Objects.requireNonNull(clazz);
+    this.clazz = clazz;
+  }
+
+  public static <T> ClassAwareKryoCoder<T> of(Class<T> clazz) {
+    return new ClassAwareKryoCoder<>(clazz);
+  }
+
+  @Override
+  public synchronized void encode(T value, OutputStream outStream)
+      throws CoderException, IOException {
+    Output output = getOutput();
+    output.setOutputStream(outStream);
+    getKryo().writeObjectOrNull(output, value, clazz);
+    output.flush();
+  }
+
+  @Override
+  public synchronized T decode(InputStream inStream) throws CoderException, IOException {
```

Comment:
i think it should be enough to synchronize `getKryo` method and have a pool of reusable input buffers, this may become a bottleneck in case of multiple tasks using the same coder